Implement scaling support for global AD indexing

lindsayad opened the pull request:
Also make it so that if a user requests global indexing in their `configure`, then they get it for finite elements now too.

- [x] Add manual scaling test

Closes #15732

lindsayad committed:
Implement scaling support when using global AD indexing

Closes #15732

lindsayad committed:
Update all framework AD compute objects to leverage global indexing

lindsayad committed:
Update navier stokes AD objects for global indexing

lindsayad committed:
Make local indexing work again for ADDG/InterfaceKernel

We needed to make `compute(OffDiag)ElemNeighJacobian` methods work
with the same number of calls for both local and global indexing. This
is a little tricky. It involved only calling those methods with values
of `Moose::ElementElement` and `Moose::NeighborNeighbor` in order to
make global indexing work...so then for local indexing to work those
methods have to individually handle `Moose::ElementNeighbor` and
`Moose::NeighborElement`

lindsayad committed:
Protect against variables not on our current subdomain in ADKernel

lindsayad commented:
@joshuahansel

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/605853/) on a2dcd03 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16026/site)

This comment will be updated on new commits.

lindsayad committed:
Handle rare bug with local AD indexing and displaced simulations

Per my comment in the source code:

When we have a displaced problem and we have AD objects it's possible that you can have
something like the following: A displaced displacement kernel uses a material property
computed with an undisplaced material. That material property is a function of the
temperature. However, the temperature doesn't have any displaced kernels acting on it nor has
any displaced objects explicitly coupling it. In this case the displaced temperature would
not register as an active_elemental_moose_variable, and if we only prepare
active_elemental_moose_variables then in our ADKernel (when not using global AD indexing) we
will either have to skip over variables who have no dof indices or we will attempt to index
out of bounds into _local_ke. So we need to make sure here that we prepare all variables that
are active *either* in the undisplaced or displaced system

lindsayad committed:
Protect against coupling corner cases in ADIntegratedBC/ADDGKernel

lindsayad committed:
Global indexing support for displaced calculations in Assembly

lindsayad committed:
Global indexing support for MooseVariableScalar

lindsayad committed:
Global indexing support in ADComputeIncrementalShellStrain

lindsayad committed:
Slightly perturbed contact matrix with global indexing more robust with NEWTON

lindsayad committed:
Dont set zeros in matrix from nodal bcs

a test case where two variables are not truly coupled. Global AD
indexing only adds into Jacobian during kernel evaluation if there is a
non-zero dependence. Then we close before nodal bcs, and the matrix
shrinks, eliminating all the off-diagional memory. Then this stupid
NodalBC::offDiagJacobian tries to set matrix value for all jvar even if
its zero, causing a new nonzero

lindsayad committed:
Make global indexing work with hanging nodes, e.g. libMesh constraints

lindsayad committed:
More scaling tests

lindsayad committed:
Ignore svg files

rwcarlsen commented on framework/src/interfacekernels/ADInterfaceKernel.C:
```diff
@@ -149,23 +152,48 @@ ADInterfaceKernelTempl<T>::computeElemNeighJacobian(Moose::DGJacobianType type)
+  auto local_functor = [&](const std::vector<ADReal> & input_residuals,
+                           const std::vector<dof_id_type> &,
+                           const std::set<TagID> &) {
+    auto i_heard_you_liked_lambdas = [&](const Moose::DGJacobianType nested_type) {
```
is that a lambda in your pocket?

rwcarlsen commented on framework/src/kernels/ADKernel.C:
```diff
@@ -138,36 +138,63 @@ ADKernelTempl<T>::computeResidual()
-        for (_j = 0; _j < _var.phiSize(); _j++)
-          _local_ke(_i, _j) += residual.derivatives()[ad_offset + _j];
-      }
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      _r = _ad_JxW[_qp];
+      _r *= _ad_coord[_qp];
+      for (_i = 0; _i < _test.size(); _i++)
+        _residuals[_i] += _r * computeQpResidual();
+    }
```
What is the purpose for collecting the separated residual contribs here instead of just dumping them into local_ke like before?

rwcarlsen replied:
Ah - I see - this is how you add the per-entry scaling - which wouldn't work before when they were all munged together.  Is there anywhere in doc comments or commit messages that explains this necessity/reason?  That might be worth capturing for future poor souls who need to dig around in this code.

lindsayad replied:
This is somewhat related to global indexing, but this code was like this well before. Even for non-global you want to save off the residuals because you can use those residuals to fill `_local_ke` for every associated jvariable. We do not want to re-do these calculations for every jvariable and corresponding `_local_ke`

lindsayad replied:
I added more doxygen to `computeResidualsForJacobian` in `ADKernel.h` and `ADIntegratedBC.h` indicating the purpose of the method

rwcarlsen commented on framework/src/kernels/ADKernel.C:
```diff
@@ -138,36 +138,63 @@ ADKernelTempl<T>::computeResidual()
   if (_has_diag_save_in && !_sys.computingScalingJacobian())
   {
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+    mooseError("_local_ke not computed for global AD indexing. Save-in is deprecated anyways. Use "
```
"anyway"

lindsayad replied:
:+1:

rwcarlsen commented on framework/src/kernels/ADKernel.C:
```diff
@@ -138,36 +138,63 @@ ADKernelTempl<T>::computeResidual()
 
   if (_has_diag_save_in && !_sys.computingScalingJacobian())
   {
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```
same here - it would be nice to keep this logic out of kernels, bcs, etc.  But maybe it's not worth the effort for such a low-level class like ADKernel.

rwcarlsen reviewed (changes requested):
It looks mostly good.  Now what we need to do is make it fast enough that we can just delete all that local indexing code nonsense.

lindsayad commented on framework/src/dgkernels/ADDGKernel.C:
```diff
@@ -139,35 +139,74 @@ ADDGKernel::computeElemNeighResidual(Moose::DGResidualType type)
   for (_qp = 0; _qp < _qrule->n_points(); _qp++)
     for (_i = 0; _i < test_space.size(); _i++)
+      residuals[_i] +=
+          _JxW[_qp] * _coord[_qp] *
```
This would save some `Real` math but wouldn't save any `ADReal` math since `computeQpResidual()` has to happen in in the inner loop. Keeping it this way keeps it consistent with what we do elsewhere (except when doing displaced calculations in `ADKernel`)

lindsayad commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -129,19 +140,27 @@ ADNodalBCTempl<T>::computeOffDiagJacobian(unsigned int jvar)
+                                                  cached_col,
+                                                  conversionHelper(residual, i)
+                                                      .derivatives()[
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```
I don't know how to do that without totally changing this code. This is telling `Assembly` to cache the value for a given row and column index; we need to provide the value. We could add a pretty dumb looking utility like this:
```
namespace Moose
{
#ifdef MOOSE_GLOBAL_AD_INDEXING
std::size_t
derivativeIndex(const std::size_t global_index, const std::size_t)
{
  return global_index;
}
#else
std::size_t
derivativeIndex(const std::size_t, const std::size_t local_index)
{
  return local_index;
}
#endif
}
```
and then we can call `Moose::derivativeIndex` here instead of having the `#ifdefs`.

lindsayad replied:
But then someone who comes and overrides `computeOffDiagJacobian` would also remember to call that utility. My solution would be to make `computeOffDiagJacobian` and `computeJacobian` `final` for `ADNodalBC`

lindsayad replied:
So I went ahead and made all `computeJacobian` related code in all AD objects both `private` and `final` so they can neither be called nor overridden. `computeResidual` and `computeResidualsForJacobian` are still overridable for `ADKernel` and `ADIntegratedBC` derived objects since some users appear to want to override those (see the AD.*Shell stuff in tensor mechanics). Let me know if you still don't like this (but know that I like it haha)

lindsayad committed:
Address review

lindsayad commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -114,7 +123,9 @@ ADNodalBCTempl<T>::computeOffDiagJacobian(unsigned int jvar)
   else
   {
+#ifndef MOOSE_GLOBAL_AD_INDEXING
     auto ad_offset = jvar * _sys.getMaxVarNDofsPerNode();
```
:+1:

lindsayad commented on framework/src/bcs/ADIntegratedBC.C:
```diff
@@ -117,85 +117,107 @@ ADIntegratedBCTempl<T>::computeResidual()
 template <typename T>
 void
-ADIntegratedBCTempl<T>::computeJacobianBlock(MooseVariableFEBase & jvar)
+ADIntegratedBCTempl<T>::computeADOffDiagJacobian()
```
I created a `computeADJacobian` method that both `computeJacobian` and `computeOffDiagJacobian` call. Lemme know if that looks good

lindsayad commented on framework/src/bcs/ADIntegratedBC.C:
```diff
@@ -117,85 +117,107 @@ ADIntegratedBCTempl<T>::computeResidual()
   {
-    unsigned int rows = ke.m();
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+    mooseError("_local_ke not computed for global AD indexing. Save-in is deprecated anyways. Use "
```
:+1:

lindsayad committed:
Allow user flexibility to define dof indices

Just as we allow someone deriving from `ADKernel` to modify the
`_residuals` data member in their `computeResidualsForJacobian`
overrides, we must also potentially allow them to modify the dof
indices. E.g. a user (it's me) has an `LMKernel` which sums computed
strong residuals into both primal and LM residuals. Hence I need to be
able to the dof indices fed into `Assembly::processDerivatives` to
include both the primal and LM degrees of freedom
